packages/components/src/range-control/index.tsx

```diff
@@ -35,67 +35,79 @@ import {
 	Wrapper,
 } from './styles/range-control-styles';
 
+import type { RangeControlProps } from './types';
+import type { WordPressComponentProps } from '../ui/context';
+
 const noop = () => {};
 
-function RangeControl(
-	{
+function UnforwardedRangeControl< IconProps = unknown >(
+	props: WordPressComponentProps<
+		RangeControlProps< IconProps >,
+		'div',
+		false
+	>,
+	ref: ForwardedRef< HTMLDivElement >
```

Comment:
Further down the code, It looks like `ref` is being passed an `HTMLInputElement`:

https://github.com/WordPress/gutenberg/blob/59f61d0689e1c00114fb724fc3f0a9c0802af90b/packages/components/src/range-control/index.tsx#L107-L113

Therefore, shouldn't `ref` be typed as a `ForwardedRef< HTMLInputElement >` too? I'm personally a bit confused to why TypeScript is not showing an error here

Reply:
Wait, isn't it supposed to be this 👇 to begin with?

```suggestion
	ref: ForwardedRef< HTMLInputElement >
```

---

> I'm personally a bit confused to why TypeScript is not showing an error here

```ts
type ForwardedRef<T> = ((instance: T | null) => void) | MutableRefObject<T | null> | null;

interface MutableRefObject<T> {
  current: T;
}
```

I guess it _is_ valid if it matches against `MutableRefObject< HTMLDivElement>` (instead of `(instance: HTMLDivElement) => void`)?

```js
// 👍 valid
const whateverFunc: React.MutableRefObject< HTMLDivElement > = ( myInput: HTMLInputElement ) => {};
```

If so, that kind of means the `React.ForwardedRef<T>` generic is mostly for type safety of methods on `ref.current` (e.g. `ref.current.focus()`), and not particularly useful for checking arg types in `ref(foo)` patterns.